### PR TITLE
DTSRD-3529. According to the changed requirement.

### PR DESCRIPTION
**Pull request by SabinaHMCTS**
**Before creating a pull request make sure that:**

- [ ] commit messages are meaningful and follow good commit message guidelines
- [ ] README and other documentation has been updated / added (if needed)
- [ ] tests have been updated / new tests has been added (if needed)

Please remove this line and everything above and fill the following sections:

### Change description ###

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[ ] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/rd/commondata/camel/processor/CategoriesProcessor.java [hunk 1]
```diff
@@ -67,18 +79,40 @@ public void process(Exchange exchange) {
...
+        if (!zeroByteCharacterRecords.isEmpty()) {
+            List<Pair<String, Long>> distinctZeroByteCharacterRecords = zeroByteCharacterRecords.stream()
+                .distinct().toList();
+            audit(distinctZeroByteCharacterRecords, null, exchange, ZERO_BYTE_CHARACTER_ERROR_MESSAGE);
+        }
+        exchange.getMessage().setBody(finalCategoriesList);
     }
 
+    public void audit(List<Pair<String, Long>> invalidCategoryIds,
+                          String fieldError,Exchange exchange,String message) {
```
Comment (SabinaHMCTS) [inline src/main/java/uk/gov/hmcts/reform/rd/commondata/camel/processor/CategoriesProcessor.java:106]: this is the change ..only this method changed .. fetchExistingRecords method was deleted

File: src/main/java/uk/gov/hmcts/reform/rd/commondata/camel/processor/CategoriesProcessor.java [hunk 2]
```diff
@@ -97,6 +131,29 @@ private void processException(Exchange exchange,
...
+                && (
+                (!category.getExternalReference().isEmpty()
+                && category.getExternalReferenceType().isEmpty())
+                || (category.getExternalReference().isEmpty()
+                    && !category.getExternalReferenceType().isEmpty())
+                )) {
+                invalidCategories.add(category);
+            }
+        }
+        finalCategoriesList.removeAll(invalidCategories);
```
Comment (SabinaHMCTS) [inline src/main/java/uk/gov/hmcts/reform/rd/commondata/camel/processor/CategoriesProcessor.java:147]: this is the actual change ... removing faulty records if any from final list and still loading file

File: src/functionalTest/resources/sourceFiles/otherCategories/other_categories_0_byte_character.csv [hunk 3]
```diff
@@ -1,4 +1,4 @@
-categorykey,serviceid,key,value_en,value_cy,hinttext_en,hinttext_cy,lov_order,parentcategory,parentkey,active
-AdditionalFacilities,,AF-VF,Video  Facility,,,,,,,Y
-AdditionalFacilities,,AF-SSC​,Same Sex Courtroom ,,,,,,,Y
-AdditionalFacilities,,AF-WR,Witness  Room,,,,,,,Y
+categorykey,serviceid,key,value_en,value_cy,hinttext_en,hinttext_cy,lov_order,parentcategory,parentkey,active,external_reference,external_reference_type
```
Comment (mikebrownccd) [inline src/functionalTest/resources/sourceFiles/otherCategories/other_categories_0_byte_character.csv:1]: 2 commas added to header, 3 commas added to data rows. maybe on purpose

File: src/main/java/uk/gov/hmcts/reform/rd/commondata/camel/mapper/CategoriesMapper.java [hunk 4]
```diff
@@ -16,6 +16,7 @@ public Map<String, Object> getMap(Object categoriesObj) {
         return CommonMapper.getMap(categories.getCategoryKey(), categories.getServiceId(), categories.getKey(),
             categories.getValueEN(), categories.getValueCY(), categories.getHintTextEN(), categories.getHintTextCY(),
             categories.getParentCategory(), categories.getParentKey(), categories.getActive(),
-            categories.getLovOrder());
+            categories.getExternalReference(),
```
Comment (mikebrownccd) [inline src/main/java/uk/gov/hmcts/reform/rd/commondata/camel/mapper/CategoriesMapper.java:19]: maybe list can be reformatted

File: src/main/java/uk/gov/hmcts/reform/rd/commondata/camel/mapper/CommonMapper.java [hunk 5]
```diff
@@ -14,7 +14,8 @@ private CommonMapper() {
 
     public static Map<String, Object> getMap(String categoryKey, String serviceId, String key, String valueEn,
                                       String valueCy, String hintTextEn, String hintTextCy, String parentCategory,
-                                      String parentKey, String active, String lovOrder) {
+                                      String parentKey, String active, String externalReference,
+                                             String externalReferenceType, String lovOrder) {
```
Comment (mikebrownccd) [inline src/main/java/uk/gov/hmcts/reform/rd/commondata/camel/mapper/CommonMapper.java:18]: params should align

File: src/main/java/uk/gov/hmcts/reform/rd/commondata/camel/mapper/OtherCategoriesMapper.java [hunk 6]
```diff
@@ -16,7 +16,8 @@ public Map<String, Object> getMap(Object categoriesObj) {
         return CommonMapper.getMap(categories.getCategoryKey(), categories.getServiceId(), categories.getKey(),
             categories.getValueEN(), categories.getValueCY(), categories.getHintTextEN(),
             categories.getHintTextCY(), categories.getParentCategory(),
-            categories.getParentKey(), categories.getActive(), categories.getLovOrder());
+            categories.getParentKey(), categories.getActive(), categories.getExternalReference(),
+                                   categories.getExternalReferenceType(),categories.getLovOrder());
```
Comment (mikebrownccd) [inline src/main/java/uk/gov/hmcts/reform/rd/commondata/camel/mapper/OtherCategoriesMapper.java:20]: list can be reformatted

File: src/main/resources/db/migration/V1_1__init_tables.sql [hunk 7]
```diff
@@ -1777,7 +1777,7 @@ INSERT INTO entity_role_codes (categorykey,serviceid,"key",value_en,value_cy,hin
  Active varchar(1)
  );
 
- INSERT INTO public.panel_member_type (categorykey,serviceid,"key",value_en,value_cy,hinttext_en,hinttext_cy,lov_order,parentcategory,parentkey,active) VALUES
+ INSERT INTO dbcommondata.panel_member_type (categorykey,serviceid,"key",value_en,value_cy,hinttext_en,hinttext_cy,lov_order,parentcategory,parentkey,active) VALUES
```
Comment (mikebrownccd) [inline src/main/resources/db/migration/V1_1__init_tables.sql:1780]: i don't think flyway likes the already applied deltas to be modified

File: src/functionalTest/resources/sourceFiles/otherCategories/other_categories_0_byte_character.csv (hunk 3 above)
Comment (SabinaHMCTS) [reply src/functionalTest/resources/sourceFiles/otherCategories/other_categories_0_byte_character.csv:1]: done

File: src/main/java/uk/gov/hmcts/reform/rd/commondata/camel/mapper/CategoriesMapper.java (hunk 4 above)
Comment (SabinaHMCTS) [reply src/main/java/uk/gov/hmcts/reform/rd/commondata/camel/mapper/CategoriesMapper.java:19]: its auto aligned by IDE

File: src/main/java/uk/gov/hmcts/reform/rd/commondata/camel/mapper/OtherCategoriesMapper.java (hunk 6 above)
Comment (SabinaHMCTS) [reply src/main/java/uk/gov/hmcts/reform/rd/commondata/camel/mapper/OtherCategoriesMapper.java:20]: its auto aligned by IDE

File: src/main/resources/db/migration/V1_1__init_tables.sql (hunk 7 above)
Comment (SabinaHMCTS) [reply src/main/resources/db/migration/V1_1__init_tables.sql]: removed change should not have been checked in

**Review by mikebrownccd — APPROVED** [review]
LGTM

**Review by lukasz-wolski — CHANGES_REQUESTED** [review]
Test are nicely formatted.
Implementation would benefit from same level of attention.

Please see comments inlined.

File: src/main/java/uk/gov/hmcts/reform/rd/commondata/configuration/DataQualityCheckConfiguration.java [hunk 8]
```diff
@@ -66,8 +46,6 @@ public <T> void processExceptionRecords(Exchange exchange,
                         ));
                     }
                 }));
-            audit(zeroByteCharacterRecords,exchange,applicationContext,flagServiceJsrValidatorInitializer);
-
         } else if (((List) typeList.get(0)).get(0).getClass().getName().contains("FlagDetails")) {
             List<FlagDetails> flagDetailList = (List<FlagDetails>) typeList.get(0);
             JsrValidatorInitializer<FlagDetails> flagDetailsJsrValidatorInitializer =
```
[the remark this answers is not in the document]
Comment (SabinaHMCTS) [reply src/main/java/uk/gov/hmcts/reform/rd/commondata/configuration/DataQualityCheckConfiguration.java]: removed

File: src/main/java/uk/gov/hmcts/reform/rd/commondata/camel/processor/CategoriesProcessor.java [hunk 9]
```diff
@@ -97,6 +131,29 @@ private void processException(Exchange exchange,
...
 
+    private List<Categories> validateExternalReference(List<Categories> finalCategoriesList, Exchange exchange) {
+        List<Categories> invalidCategories = new LinkedList<>();
+        for (Categories category : finalCategoriesList) {
+            if ((category.getExternalReference() != null && category.getExternalReferenceType() != null)
+                && (
+                (!category.getExternalReference().isEmpty()
+                && category.getExternalReferenceType().isEmpty())
+                || (category.getExternalReference().isEmpty()
+                    && !category.getExternalReferenceType().isEmpty())
```
[the remark this answers is not in the document]
Comment (SabinaHMCTS) [reply src/main/java/uk/gov/hmcts/reform/rd/commondata/camel/processor/CategoriesProcessor.java]: done

File: src/main/java/uk/gov/hmcts/reform/rd/commondata/camel/mapper/CategoriesMapper.java [hunk 10]
```diff
@@ -16,6 +16,7 @@ public Map<String, Object> getMap(Object categoriesObj) {
...
+                                   categories.getExternalReferenceType(), categories.getLovOrder());
```
[the remark this answers is not in the document]
Comment (SabinaHMCTS) [reply src/main/java/uk/gov/hmcts/reform/rd/commondata/camel/mapper/CategoriesMapper.java]: done

**Review by lukasz-wolski — CHANGES_REQUESTED** [review]
please see comment inlined

File: src/main/java/uk/gov/hmcts/reform/rd/commondata/configuration/DataQualityCheckConfiguration.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (lukasz-wolski) [reply src/main/java/uk/gov/hmcts/reform/rd/commondata/configuration/DataQualityCheckConfiguration.java]: 'validator' method parameter no longer used by method implementation, please remove it

File: src/functionalTest/java/uk/gov/hmcts/reform/rd/commondata/cameltest/CommonDataCategoriesLoadTest.java [hunk 11]
```diff
@@ -313,7 +313,71 @@ void testListOfValuesCsv_0_byte_character() throws Exception {
...
+        Pair<String, String> pair = new Pair<>(
+            UPLOAD_LIST_OF_VALUES_FILE_NAME,
+            externalReferenceErrorMessage
+        );
+        validateCategoriesFileException(jdbcTemplate, exceptionQuery, pair);
+        validateCategoriesFileAudit(
+            jdbcTemplate,
+            auditSchedulerQuery,
+            "PartialSuccess",
             UPLOAD_LIST_OF_VALUES_FILE_NAME
```
Comment (rajkatla-hmcts) [inline src/functionalTest/java/uk/gov/hmcts/reform/rd/commondata/cameltest/CommonDataCategoriesLoadTest.java:381]: Create test cases to cover below scenarios, please :

1. All records contain null external_reference_type and null external_reference
2. Missing new headers

**Review by rajkatla-hmcts — CHANGES_REQUESTED** [review]

File: src/main/java/uk/gov/hmcts/reform/rd/commondata/camel/processor/CategoriesProcessor.java [hunk 12]
```diff
@@ -38,9 +38,21 @@ public class CategoriesProcessor extends JsrValidationBaseProcessor<Categories>
...
+
+
     public static final String LOV_COMPOSITE_KEY = "categorykey,key,serviceid";
+
+    public static final String LOV_EXTERNAL_REFERENCE = "external_reference,external_reference_type";
+
     public static final String LOV_COMPOSITE_KEY_ERROR_MSG = "Composite Key violation";
 
+    public static final String EXTERNAL_REFERENCE_ERROR_MSG = "Both external_reference and "
+        + "external_reference_type value must be null or both must be not-null";
```
[the remark this answers is not in the document]
Comment (SabinaHMCTS) [reply src/main/java/uk/gov/hmcts/reform/rd/commondata/camel/processor/CategoriesProcessor.java:50]: any suggestion

File: src/main/java/uk/gov/hmcts/reform/rd/commondata/configuration/DataQualityCheckConfiguration.java [hunk 13]
```diff
@@ -13,34 +11,21 @@
...
-        "Zero byte characters identified - check source file";
-
     @Value("${zero-byte-characters}")
     public List<String> zeroByteCharacters;
 
-    public <T> void processExceptionRecords(Exchange exchange,
-                                            List<Object> typeList,
-                                            ApplicationContext applicationContext,
+    public <T> List<Pair<String, Long>> processExceptionRecords(List<Object> typeList,
                                             JsrValidatorInitializer<T> validator) {
```
[the remark this answers is not in the document]
Comment (SabinaHMCTS) [reply src/main/java/uk/gov/hmcts/reform/rd/commondata/configuration/DataQualityCheckConfiguration.java]: removed

File: src/functionalTest/java/uk/gov/hmcts/reform/rd/commondata/cameltest/CommonDataCategoriesLoadTest.java (hunk 11 above)
Comment (rajkatla-hmcts) [reply src/functionalTest/java/uk/gov/hmcts/reform/rd/commondata/cameltest/CommonDataCategoriesLoadTest.java:381]: @SabinaHMCTS as discussed , ignore the missing header scenario for now as you will need more time for that, i will create a separate ticket to cover that and for now just cover 1'st scenario

Comment (SabinaHMCTS) [reply src/functionalTest/java/uk/gov/hmcts/reform/rd/commondata/cameltest/CommonDataCategoriesLoadTest.java:381]: thank you. for 1.All records contain null external_reference_type and null external_reference .. do we need a test case success scenerio - upload without external reference? and show that the values are null?

Comment (SabinaHMCTS) [reply src/functionalTest/java/uk/gov/hmcts/reform/rd/commondata/cameltest/CommonDataCategoriesLoadTest.java:422]: added testListOfValuesAllRecordsContainNullExternalReferenceTypeAndNullExternalReferenceSuccess

**Review by lukasz-wolski — APPROVED** [review]
please see suggestion inlined

File: src/functionalTest/java/uk/gov/hmcts/reform/rd/commondata/cameltest/CommonDataCategoriesLoadTest.java [hunk 14]
```diff
@@ -313,7 +314,111 @@ void testListOfValuesCsv_0_byte_character() throws Exception {
...
+        jobLauncherTestUtils.launchJob();
+        var listOfValues = jdbcTemplate.queryForList(listOfValuesSelectData);
+        assertEquals(3, listOfValues.size());
+        //Validate Success Result
+        assertThat(listOfValues.get(0).get("categoryKey")).isEqualTo("panelCategoryMember");
+        assertThat(listOfValues.get(0).get("key")).isEqualTo("PC1-01-74");
+        assertThat(listOfValues.get(0).get("parentcategory")).isEqualTo("caseSubType");
+        assertThat(listOfValues.get(0).get("parentkey")).isEqualTo("PC2");
+        assertThat(listOfValues.get(0).get("external_reference")).isEqualTo("");
+        assertThat(listOfValues.get(0).get("external_reference_type")).isEqualTo("");
```
[the remark this answers is not in the document]
Comment (SabinaHMCTS) [reply src/functionalTest/java/uk/gov/hmcts/reform/rd/commondata/cameltest/CommonDataCategoriesLoadTest.java:375]: Postgress treats empty values as empty and not null unless its explicitly sent as null, unlike Oracle which will convert .
for lov , if the csv has sent empty values its saves as empty , but fro lov order before we save file there is a chack in code / mapper
if (isBlank(lovOrder)) {
            categoriesRow.put("lov_order", null);
        } else {
            categoriesRow.put("lov_order", Long.valueOf(trim(lovOrder)));
        }
so we explicitly add null in the db so thsts the only column. witjh null values

**Review by rajkatla-hmcts — APPROVED** [review]
